<syncstream>: implement basic_syncbuf and basic_osyncstream

MichaelRizkalla opened the pull request:
This PR will implement [P0053R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0053r7.pdf) and [P0753R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0753r2.pdf). should close #3.

It's still on going work, current progress tracking:
- [x] Implement `basic_syncbuf`
- [x] Implement `basic_osyncstream`
- [x] Implement `osyncstream` manipulators
- [x] Implement tests for `basic_syncbuf`
- [x] Implement tests for `basic_osyncstream`
- [x] Implement tests for `osyncstream` manipulators

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};
```
@miscco, you were saying that this could be `default`. I understand that it is required to initialize the allocator using `_Alloc{}` for a default constructed `basic_syncbuf`.
Can this be achieved using member initialization?
```C++
basic_syncbuf() = default;
// ...
streambuf_type* _Wrapped = nullptr;
allocator_type _Al = _Alloc{};
bool _Emit_on_sync = false;
bool _Record_sync  = false;
```

StephanTLavavej replied:
You can just say `allocator_type _Al{};` as either brace-initialization or copy-initialization (with `=`) are acceptable syntax for data member initializers.

MichaelRizkalla commented:
@StephanTLavavej, since this is a new header, can you please look into adding the relevant changes to the internal repo if required?

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+            lock_guard<mutex> _Guard(_Mutex);
+
+            auto& _Mutex_map = _Get_mutex_map();
+            _Mutex_map.emplace(std::piecewise_construct, std::forward_as_tuple(_Ptr), std::forward_as_tuple())
+                .first->second._Ref_count++;
```
```suggestion
            _Mutex_map.try_emplace(_Ptr).first->second._Ref_count++;
```

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+struct _Basic_syncbuf_global_mutex {
+private:
+    struct _Mutex_count_pair {
+        mutex _Mutex;
+        uint64_t _Ref_count;
+
+        _Mutex_count_pair() : _Ref_count(0), _Mutex(mutex{}) {}
+        ~_Mutex_count_pair()                        = default;
+        _Mutex_count_pair(_Mutex_count_pair&&)      = delete;
+        _Mutex_count_pair(const _Mutex_count_pair&) = delete;
+        _Mutex_count_pair& operator=(const _Mutex_count_pair&) = delete;
+        _Mutex_count_pair& operator=(_Mutex_count_pair&&) = delete;
```
```suggestion
        mutex _Mutex;
        uint64_t _Ref_count = 0;
```
No need for all of these members

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+    static mutex& _Get_mutex_for_instance(void* _Ptr) {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
```
```suggestion
            scoped_lock _Guard(_Mutex);
```

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+    static void _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
```
```suggestion
            scoped_lock _Guard(_Mutex);
```

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+    static void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
```
```suggestion
            scoped_lock _Guard(_Mutex);
```

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            lock_guard<mutex> _Guard(
```
```suggestion
            scoped_lock _Guard(
```

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+};
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
```
```suggestion
template <class _Elem, class _Traits, class _Alloc>
```
You are redefining default parameters and it's causing tests to fail.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,330 @@
+        return _Mutex;
+    }
+    static mutex& _Get_mutex_for_instance(void* _Ptr) {
+        mutex& _Mutex = _Get_map_mutex();
```
This variable is only used to initialize `scoped_lock`. It should be inlined. Ditto below.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,330 @@
+struct _Basic_syncbuf_global_mutex {
+private:
+    struct _Mutex_count_pair {
+        mutex _Mutex;
```
As I said previously, we should use `shared_mutex`.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,330 @@
+#undef emit
+
+_STD_BEGIN
+struct _Basic_syncbuf_global_mutex {
```
Why making this struct?

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,330 @@
+        uint64_t _Ref_count = 0;
+    };
+    static map<void*, _Mutex_count_pair>& _Get_mutex_map() noexcept {
+        static map<void*, _Mutex_count_pair> _Mutex_Map;
```
We shouldn't use function scope static variables, so this and map mutex should be brought to namespace scope (after that these functions can be removed). They might have to be built into a DLL but I'm not sure about that.

MichaelRizkalla replied:
I get from your review, that we should allow these functions to exist directly in `::std` namespace. And to achieve thread-safety, the map could be in a DLL. I am wondering how I can add the variables in the DLL "not sure how it works for STL 😕"

AdamBucior replied:
Unfortunately I also don't know it so we need to ask a maintainer.

MichaelRizkalla reviewed (commented):
I started this review to have all questions that I think need a maintainer's input.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
```
Fortunately, this is unnecessary. `/clr:pure` and plain `/clr` are incompatible with `/std:c++latest`, so we don't need to worry about them for C++20-only features:

```
C:\Temp>cl /clr:pure /nologo /W4 /std:c++latest /c meow.cpp
cl : Command line warning D9035 : option 'clr:pure' has been deprecated and will be removed in a future release
meow.cpp
c1xx: warning C4857: C++/CLI mode does not support C++ versions newer than C++17; setting language to /std:c++17
c1xx: error C2338: two-phase name lookup is not supported for C++/CLI or C++/CX; use /Zc:twoPhase-
c1xx: fatal error C1903: unable to recover from previous error(s); stopping compilation

C:\Temp>cl /clr /nologo /W4 /std:c++latest /c meow.cpp
meow.cpp
c1xx: warning C4857: C++/CLI mode does not support C++ versions newer than C++17; setting language to /std:c++17
c1xx: error C2338: two-phase name lookup is not supported for C++/CLI or C++/CX; use /Zc:twoPhase-
c1xx: fatal error C1903: unable to recover from previous error(s); stopping compilation
```

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,26 @@
+
+#include <syncstream>
+
+#ifdef __cpp_lib_syncbuf
```
As the STL's separately compiled DLLs/LIBs (implemented in `stl/src`) are built by our build system with a single compiler (MSVC) with a single Standard mode (`/std:c++latest`, except for the managed code stuff), we don't have to worry about features being conditionally present.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+    uint64_t _Ref_count = 0;
+};
+
+_CRTIMP2_PURE map<void*, _Mutex_count_pair>& __cdecl _Get_mutex_map();
```
Functions that need to worry about `/clr:pure` are typically `_CRTIMP2_PURE ReturnType __CLRCALL_PURE_OR_CDECL`. I see that some of our multithreading functions are `_CRTIMP2_PURE ReturnType __cdecl` but this is probably a (cosmetic) mistake as the multithreading code typically forbids `/clr:pure`; they should probably be `_CRTIMP2 ReturnType __cdecl`. Our new code prefers flat C exports with `__stdcall` (I believe because it's marginally more efficient); `xfilesystem_abi.h` and `stl/src/atomic_wait.cpp` are examples.

However, before we get to the question of exporting and calling conventions, there's the ABI question. For somewhat arcane binary compatibility reasons, we currently avoid *adding* exports to the STL's DLL (much less modifying/removing exports which outright breaks bincompat). When we need to add separately compiled machinery, our first preference is to "inject" it into the DLL's import LIB, which effectively statically links it; this is the approach we took with Filesystem. When we additionally need global data (as is the case for memory resources, codecvt IDs, and atomic waits, as I understand it), the import LIB approach isn't sufficient, and we need to go further and add a "satellite DLL" - most recently done for atomic waits.

If we need global mutexes (I am not yet up to speed on syncstream implementation details), then the satellite DLL approach may be necessary. This is a surprisingly enormous headache (beyond just the build system changes for GitHub, there are MSVC-internal build changes *and* there's a bunch of poorly understood MSVC-internal changes that need to happen) but there is no other choice than waiting for vNext which we have tried to avoid for new features (successfully so far).

AdamBucior replied:
@StephanTLavavej Can it be added to atomic_wait DLL since c++latest has unstable ABI?

MichaelRizkalla replied:
As for now, I added the source file to build with the `IMPLIB_SOURCES` group, which I suppose is the static import library.

I am experimenting with building it with other targets, but it fails to link with `shared_mutex::lock` and `shared_mutex::unlock`. I'll let you decide the next steps since I have little experience with STL and CRT interactions.

StephanTLavavej replied:
That's an excellent idea, @AdamBucior. Aside from making the name incomplete, I see no reason why we couldn't do that, and the name being incompletely descriptive is much less of a headache for us *and* users than adding a new satellite.

AlexGuteniev replied:
C+17 parallel algorithms rely on atomic wait DLL.
I still see no issues as any version of atomic wait dll would have support for parallel algorithms.

StephanTLavavej commented:
> @StephanTLavavej, since this is a new header, can you please look into adding the relevant changes to the internal repo if required?

Yes, we'll take care of all of the internal changes mentioned in https://github.com/microsoft/STL/wiki/Files-To-Edit-When-Adding-Or-Removing-Files , when mirroring this PR. We generally implement this after the GitHub PR is fully reviewed.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+        _STD swap(_Wrapped, _Right._Wrapped);
+    }
+
+    bool _Emit() noexcept {
+        _TRY_BEGIN
+        return emit();
+        _CATCH_ALL
+        _CATCH_END
+    }
```
If this throws an exception, it will fall off of the end of the function without returning a `bool`.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+    }
+
+private:
+    enum { // constant for minimum buffer size
+        _MINSIZE = 32
+    };
```
In new code, we prefer to use `static constexpr` data members for such constants, and give them `_Ordinary_ugly_names` instead of `_MACRO_UGLY_NAMES`.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+    }
+
+    int_type overflow(int_type _Current_elem) override {
+        if (_Wrapped == nullptr)
+            return _Traits::eof();
```
We conventionally use braces for all control flow; occurs elsewhere. (This helps avoid mistakes during maintenance, and we feel that the consistency is worth the additional line.)

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+        return _Result;
+    }
+
+    streambuf_type* get_wrapped() const noexcept {
```
General comment: For every non-`void` function (especially public, but also `_Ugly_internal`), consider whether it meets our criteria for being marked `_NODISCARD`; we are trying to do this for all new code.

The most common criterion for marking something as nodiscard is when a function is a "pure observer" - with no side effects, the only reason to call a pure observer is to get its return value, so it would make no sense to discard it. (Less common criteria are "acquires a raw resource like `allocate()`" or "is almost always an error if discarded like `std::remove_if()`".) If a function has side effects and there's even a 10% chance that it can be reasonably called without using its return value, then we don't mark it - avoiding false positive warnings is extremely important here.

We are permitted to mark things as nodiscard beyond the occurrences of `[[nodiscard]]` depicted in the Standard.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+
+    basic_syncbuf(streambuf_type* _StrBuf, const _Alloc& _Al) : _Wrapped(_StrBuf), _Al(_Al) {
+        if (_Wrapped != nullptr) {
+            _Add_mutex_for_instance_or_increment(static_cast<void*>(_Wrapped));
```
Pointers implicitly convert to `void*` so this `static_cast<void*>` is unnecessary. Occurs below.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};
+
+    explicit basic_syncbuf(streambuf_type* _StrBuf) : basic_syncbuf(_StrBuf, _Alloc{}) {}
```
Naming nitpick: We conventionally use `_UglyCamel` case for template parameters like `_FwdIt`. For function parameters, we use `_Ugly_snake` case. Occurs below.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+
+    explicit basic_syncbuf(streambuf_type* _StrBuf) : basic_syncbuf(_StrBuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _StrBuf, const _Alloc& _Al) : _Wrapped(_StrBuf), _Al(_Al) {
```
We conventionally avoid all shadowing, including constructor parameter/data member shadowing. Here, we conventionally add a trailing underscore to the constructor parameter, taking `_Al_`.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    if (_Instance_mutex_itr == _Mutex_map.end())
+        return;
+    _Instance_mutex_itr->second._Ref_count--;
```
We conventionally prefer predecrement by default; as `->` and `.` bind most tightly, no parentheses are needed.

I see that this somewhat verbose expression is then read on the next line; using predecrement would allow this to be combined with the test against `== 0`. That's probably a clarity win. (Expressions that do a *lot* of modifications are typically found in our oldest code - we prefer to avoid that nowadays, but a "decrement and test if zero" is pretty simple.)

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+        const pointer _Old_ptr         = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        const size_type _Old_data_size = _Get_data_size();
+
+        const pointer _New_ptr = _Unfancy(_Al.allocate(_New_capacity));
```
Is this correct? `_Unfancy` returns a raw pointer, but `pointer` is the allocator's fancy pointer type. One of these seems incorrect (either we need `pointer` so we shouldn't unfancy, or we need to unfancy so we should be storing a raw pointer). Occurs below. We probably need test coverage for fancy pointers here.

MichaelRizkalla replied:
Yep, I think this is incorrect since I am using the same `pointer` type from the allocator.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,26 @@
+
+_STD_BEGIN
+// OBJECT DECLARATIONS
+static map<void*, _Mutex_count_pair> _Mutex_map{};
```
If we're going to have long-lived allocations in separately compiled code, we probably need to mark them as "CRT blocks" to prevent them from showing up as memory leaks (as viewed by the CRT leak tracking machinery). There's an existing example of this in STL code.

StephanTLavavej replied:
Here's an example:
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/src/cond.cpp#L39
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/src/cond.cpp#L52

MichaelRizkalla replied:
Thanks 😄, I was going to ask for an example.

MichaelRizkalla replied:
I have some concerns regarding this issue.

- We allocate `map<void*, _Mutex_count_pair>` as `_CRT_BLOCK`, but `map` allocates more elements as it grows. Would not this be detected as a memory leak?
- To ensure allocation/deallocation of these two variables, I assume we have to create an empty static struct which I named `_Init_global_mutex_map` on my local device. This struct will use 1 byte and will live for the entire program lifetime. Will this cause the memory leak false-positive as well? 😕

StephanTLavavej replied:
Yes, the allocations for the map nodes are the ones that need to be marked as CRT blocks. This can be done by giving the map a custom allocator.

The map itself doesn't need to be dynamically allocated; it can have static storage duration. No need to have an empty static struct, if I understand what's happening correctly.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,312 @@
+_CRTIMP2_PURE map<void*, _Mutex_count_pair>& __cdecl _Get_mutex_map();
+_CRTIMP2_PURE shared_mutex& __cdecl _Get_map_mutex();
+
+shared_mutex& _Get_mutex_for_instance(void* _Ptr) {
```
These functions need to either be made `inline` or moved to a DLL. Since we need to use DLL anyway it's probably better to move these functions there.

MichaelRizkalla replied:
I'll move these functions to `syncstream.cpp`, it seems the better solution like you said.

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,282 @@
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(_Wrapped));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
```
This line produces:
```C++
 warning C4389: '!=': signed/unsigned mismatch
```
This made me thinking should I keep `size_type` as `allocator_traits<_Alloc>::size_type` or change it to `streamsize` since `streamsize` could technically be larger than `size_t`

StephanTLavavej replied:
Not just technically - our `streamsize` is always `long long` even on 32-bit platforms:
https://github.com/microsoft/STL/blob/ac4fde764ae716512bdd67e0e9577bd53947bc3d/stl/inc/iosfwd#L49

MichaelRizkalla replied:
It's obvious then, I'll do the necessary changes to use `streamsize` as the `size_type` of `basic_syncbuf`.

MichaelRizkalla replied:
After re-consideration, using `size_type` is more plausible since the data held by the `streambuf` can not exceed `max_size` of the allocator anyway. a `static_cast` of the returned value from `sputn` should be enough.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,282 @@
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <map>
```
```suggestion
```
No need to include `<map>` anymore

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,282 @@
+        if (_Getal() == _Right._Getal()) {
+            _Move_assign(_Right, _Equal_allocators{});
+        } else {
+            const size_type _Right_Buf_size  = _Right._Get_buffer_size();
+            const size_type _Right_Data_size = _Right._Get_data_size();
```
These should be `_Right_buf_size` and `_Right_data_size`

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+        _Emit();
+    }
+
+    basic_osyncstream& operator=(basic_osyncstream&&) noexcept = default;
```
[LWG-3334](https://cplusplus.github.io/LWG/issue3334) removed the assignment operator's `Effect` and `Postconditions`, I am wondering if this means that the function should be `default` or be removed since it exists in the current working draft's header overview only.

StephanTLavavej replied:
WG21-N4878 16.3.3.4 [functions.within.classes]/1 "For the sake of exposition, Clause 17 through Clause 32 and Annex D do not describe copy/move constructors, assignment operators, or (non-virtual) destructors with the same apparent semantics as those that can be generated by default (11.4.5.3, 11.4.6, 11.4.7). It is unspecified whether the implementation provides explicit definitions for such member function signatures, or for virtual destructors that can be generated by default."

Defining it with `= default` seems reasonable.

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
```
I am having a look at [P0753R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0753r2.pdf) `osyncstream` Manipulators , which I thought to be an addition to the implementation of [P0053R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0053r7.pdf) as the name implies, however, there's a scary note 😨.

```
To work around the issue that the Allocator template argument can not be deduced, implementations can
introduce an intermediate base class to basic_osyncbuf that manages its emit_on_sync flag
```
I am thinking about it now, but apparently, it will be required to create an intermediate class that only manages `_Emit_on_sync` member variable.

- What do you think? Could there be another workaround that I have missed?
- If an intermediate class is required, is there a naming convention for such a thing?
- Is it wise for the intermediate class to manage `_Emit_on_sync` variable only without interacting with `basic_streambuf`'s members at all?

StephanTLavavej replied:
Can't imagine another workaround with equal efficiency. A base class with a virtual member function would be strictly worse.

There isn't much in the way of consistent precedent for base class naming - we have `_Vector_val` which holds its data members (erasing the allocator type), while `function` inherits from `_Func_class` (not a very descriptive name), `optional` has many base classes like `_Optional_construct_base`, etc.

`_Emit_on_sync_manager`, `_Emit_on_sync_base`, `_Basic_osyncbuf_impl`, etc. could all be plausible.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifndef __cpp_lib_syncbuf
```
This should be `#if !_HAS_CXX20`. (The comments also need updating)

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+_STD_BEGIN
+
+extern "C" shared_mutex& __cdecl _Get_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept;
+extern "C" void __cdecl _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept;
```
`_Acquire_mutex_for_instance` and `_Release_mutex_for_instance` would be better names for these functions.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+        _Mybase::set_rdbuf(_STD addressof(_Sync_buf));
+    }
+
+    ~basic_osyncstream() {
+        _Emit();
+    }
```
This destructor sould be `default` per LWG-3334

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,339 @@
+private:
+    void _Emit() {
+        _TRY_BEGIN
+        return emit();
```
```suggestion
        emit();
```
`emit` returns void so this `return` is superfluous.

AdamBucior replied:
Actually this function can be removed after the destructor is defaulted.

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,42 @@
+static shared_mutex _Mutex{};
+
+extern "C" _CRTIMP2 shared_mutex& _Get_mutex_for_instance(void* _Ptr) {
+    scoped_lock _Guard(_Mutex);
```
```suggestion
    shared_lock _Guard(_Mutex);
```
`find` and `end` are guaranteed to not modify `map` so this function can be called concurrently.

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,42 @@
+extern "C" _CRTIMP2 void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Mutex);
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    if (_Instance_mutex_itr == _Mutex_map.end()) {
+        return;
+    }
```
This should rather be an assert.

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,42 @@
+
+extern "C" _CRTIMP2 shared_mutex& _Get_mutex_for_instance(void* _Ptr) {
+    scoped_lock _Guard(_Mutex);
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
```
`iter` would be more conventional than `itr`. Ditto below

AdamBucior commented on stl/inc/iosfwd:
```diff
@@ -200,6 +200,10 @@ template <class _Elem, class _Traits = char_traits<_Elem>>
 class basic_ofstream;
 template <class _Elem, class _Traits = char_traits<_Elem>>
 class basic_fstream;
+template <class _Elem, class _Traits = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
+class basic_syncbuf;
+template <class _Elem, class _Traits = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
+class basic_osyncstream;
```
This (and the aliases below) should be guarded with `_HAS_CXX20`

miscco reviewed (commented):
There are some concerns about fancy pointers and allocators, but besides that mostly nits.

Great progress :+1:

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+    explicit basic_syncbuf(streambuf_type* _Strbuf) : basic_syncbuf(_Strbuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _Strbuf, const _Alloc& _Al_) : _Wrapped(_Strbuf), _Al(_Al_) {
+        if (_Wrapped != nullptr) {
```
I think that this check could be beneficial, in case an object was created explicitly with `nullptr`. The internal map will hold a non-necessary `mutex` for the program's lifetime since there's no way to erase it.

AdamBucior replied:
I think he meant to remove `!= nullptr` and stay with `if (_Wrapped)`

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+
+    void swap(basic_syncbuf& _Right) noexcept {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
```
I added `_Getal` at an early stage because I thought I might need to use `_Compressed_pair`. I think this helper function could be removed now, and use `_Al` directly everywhere.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+
+_STD_BEGIN
+
+extern "C" shared_mutex& __cdecl _Get_mutex_for_instance(void* _Ptr);
```
Returning `shared_mutex&` here causes clang warning `-Wreturn-type-c-linkage`. We need to either disable this warning or change the return type (making it a pointer would probably suffice but I'm not sure about that).

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {
```
`set_emit_on_sync` is modifying a member variable, it should not be right to mark as const 👀.

AdamBucior replied:
I think he meant to make `_Val` const.

MichaelRizkalla replied:
Again, I misunderstood the comment, my bad 🤣. It could be `const`, but this will result in a different signature than [P0053R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0053r7.pdf). Is it allowed to do such changes?

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+    }
+
+    void _Move_assign(basic_syncbuf&& _Right, _Equal_allocators) noexcept {
+        _Tidy();
```
I will add the check to `operator=` after `_Emit` call. It seems like the better option since `_Move_assign` could do this check twice if `_No_propagate_allocators` and `_Al == _Right._Al` are both true.

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,66 @@
+    constexpr _Crt_allocator(const _Crt_allocator<_Other>&) noexcept {}
+
+    _NODISCARD __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
+        return static_cast<_Ty*>(_calloc_crt(_Count, sizeof(_Ty)));
```
This function should probably check if `_calloc_crt` returns `nullptr` and throw `bad_alloc` if it does.

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,66 @@
+        return static_cast<_Ty*>(_calloc_crt(_Count, sizeof(_Ty)));
+    }
+
+    void deallocate(_Ty* const _Ptr, const size_t) {
```
```suggestion
    void deallocate(_Ty* const _Ptr, const size_t) noexcept {
```

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,66 @@
+static map<void*, _Mutex_count_pair, less<void*>, _Crt_allocator<pair<void* const, _Mutex_count_pair>>> _Mutex_map;
+static shared_mutex _Mutex;
+
+extern "C" _CRTIMP2 shared_mutex* _Get_mutex_for_instance(void* _Ptr) {
```
I see no reason why this functions can't be `noexcept`

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,66 @@
+    _ASSERT_EXPR(_Instance_mutex_iter != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _STD addressof(_Instance_mutex_iter->second._Mutex);
+}
+extern "C" _CRTIMP2 void _Acquire_mutex_for_instance(void* _Ptr) noexcept {
```
This function on the other hand shouldn't be `noexcept` because it might allocate memory.

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,329 @@
+        const pointer _Old_ptr         = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        const size_type _Old_data_size = _Get_data_size();
+
+        const pointer _New_ptr = _Al.allocate(_New_capacity);
```
I'd probably need some help here to reason about fancy vs non-fancy pointers and when to use them in this context. Per my understanding, `allocate` will return the fancy pointer type, which matches `pointer`. I am pretty sure that `streambuf_type` functions like `pbase` are returning non-fancy pointers, which are converted to fancy `pointer` type via `pointer_traits<pointer>::pointer_to`.
If a `_Traits::meow` function is called with both fancy pointers, would not this be fine 🤔?

If I am mistaken, there would be several occurrences to fix 🐱

MichaelRizkalla replied:
Follow up:
While writing tests, I encountered an issue when using an allocator that returns a fancy type, I'll do some changes to make sure the code is using unfancy pointers instead.
Resolving this one for now...

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+template <class _Elem, class _Traits>
+class _Basic_syncbuf_impl : public basic_streambuf<_Elem, _Traits> {
+public:
+    void _Set_emit_on_sync(const bool _Val) noexcept {
```
Rather than making another function it would be better to change this function to `set_emit_on_sync` and then put `using _Mybase::set_emit_on_sync` in `basic_syncbuf`.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+
+    streambuf_type* _Wrapped{nullptr};
+    allocator_type _Al{};
+    bool _Record_sync{false};
```
`_Record_sync` should be moved up to `_Basic_syncbuf_impl` to avoid padding. Also `_Sync_recorded` might be a better name.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
```
Why not use constructor initializer list?

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+        }
+    }
+
+    void _Assign_rv(basic_syncbuf&& _Right) {
```
This function is only used in the move constructor so it can be inlined. After it's done the `this != _STD addressof(_Right)` check and `_Tidy` call can be eliminated.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    void swap(_Basic_syncbuf_impl& _Right) {
+        if (this != _STD addressof(_Right)) {
```
This check is unnecessary, `basic_syncbuf` already does this check and swapping one bool is not expensive anyway.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,360 @@
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
```
This function is required by the standard to be `noexcept`.

MichaelRizkalla replied:
This is confusing to me, the class template does not have `noexcept` but later it's marked as `noexcept`.
Also, `_Move_assign` with `_No_propagate_allocators` could call `allocate` which might throw.

StephanTLavavej replied:
Yep, this looks like it should be reported as an LWG issue - `noexcept` appears to be incorrect in the Standard.

MichaelRizkalla replied:
I found LWG-3498 reporting this issue, but it has not been voted yet

StephanTLavavej replied:
Added comments citing LWG-3498. Thanks!

MichaelRizkalla commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.cpp:
```diff
@@ -0,0 +1,285 @@
+
+    test_syncbuf_move_swap_operations<allocator<char>>(&char_buffer);
+    test_syncbuf_move_swap_operations<fancy_ptr_allocator<char>>(&char_buffer);
+    // test_syncbuf_move_swap_operations<non_move_assignable_non_equal_allocator<char>>(&char_buffer);
```
This test is failing with `Exit Code: 3221225477`, and it's also failing locally on my local machine with MSVC.
It fails with an exception while exiting the destructor of `buf1` object in the ```// test move assignment``` section with access violation from msvcp140d.dll. I am not sure how to get to the bottom of this one, any help 😕?

MichaelRizkalla reviewed (commented):
I started sorting out the tests for this PR.
I do not really understand why `P1502R1_standard_library_header_units` is failing now, so I'd like your input.

MichaelRizkalla commented:
I _think_ the PR is ready for review. 
There's one failing test, with no apparent reason to me. I probably overlooked something that should be caught in the review 😄.

crackedmind commented:
You forgot to update header-units.json

Berrysoft commented on stl/inc/ostream:
```diff
@@ -993,6 +993,38 @@ basic_ostream<_Elem, _Traits>& __CLRCALL_OR_CDECL flush(basic_ostream<_Elem, _Tr
+
+template <class _Elem, class _Traits>
+basic_ostream<_Elem, _Traits>& noemit_on_flush(basic_ostream<_Elem, _Traits>& _Ostr) {
+    auto _Sync_buf_ptr = dynamic_cast<_Basic_syncbuf_impl<_Elem, _Traits>*>(_Ostr.rdbuf());
```
The compiler may not enable RTTI and thus no `dynamic_cast` avaliable. I've searched through the whole repo, and only 2 places using it, with `#ifdef _CPPRTTI` guard.

MichaelRizkalla replied:
Thanks, I totally overlooked that option.

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,355 @@
+            _Traits::copy(_New_ptr, _Right.pbase(), _Right_data_size);
+
+            streambuf_type::setp(_New_ptr, _New_ptr + _Right_data_size, _New_ptr + _Right_buf_size);
+            streambuf_type::_Plocale = _STD move(_Right._Plocale);
```
You are moving the pointer rather than swapping it, so it's deallocated twice. This causes `test_syncbuf_move_swap_operations<non_move_assignable_non_equal_allocator<char>>` to fail (took _only_ an hour to find).
```suggestion
            _STD swap(streambuf_type::_Plocale, _Right._Plocale);
```

MichaelRizkalla replied:
Good catch!
That's a subtle one to find, I would have never guessed 😢.

AdamBucior commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.hpp:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    void deallocate(pointer const _Ptr, const size_t _Count) noexcept {
+        ::operator delete(&*_Ptr, _Count);
```
All of the allocators are deallocating with `_Count` rather than `_Count * sizeof(value_type)` which is probably wrong.

Berrysoft replied:
I suggest using `std::allocator` for this purpose ;)

MichaelRizkalla replied:
@Berrysoft, I _suppose_ tests do not care about performance, then using `std::allocator` will be a safe option to avoid allocation mistakes like this one :cat:.

MichaelRizkalla commented on tests/std/tests/P1502R1_standard_library_header_units/test.cpp:
```diff
@@ -804,7 +804,13 @@ int main() {
+        syncbuf sync_buf{nullptr};
+        assert(sync_buf.get_wrapped() == nullptr);
+        assert(sync_buf.get_allocator() == std::allocator<char>{});
+        assert(sync_buf.emit() == false);
```
This line of code is causing the failure of `P1502R1_standard_library_header_units`. I ran it locally and commenting out this line make it pass. This function call should return `false` with no operations to be executed.

AdamBucior replied:
Have you tested if it passes without the `assert`? (only the call to emit)

MichaelRizkalla replied:
On my local machine, the call itself makes the test fails.
```C++
assert(sync_buf.emit() == false); // fails, exit code: 3221225477
assert(sync_buf.emit() == true); // fails. Assertion failed ... line 810
sync_buf.emit(); // fails, exit code: 3221225477
```

MichaelRizkalla replied:
Follow up:
I found out that this failure happens when using `/MTd` compiler flag. I am still not sure what this means so I'll update the comment if I found anything relevant :cat:

StephanTLavavej replied:
I can reduce and report this compiler bug.

StephanTLavavej replied:
Reported internally as VSO-1275701 "Standard Library Header Units: Runtime crash with dynamic atexit destructor for syncstreams". I'll see if it continues to repro after the satellite DLL change; if it does, I'll disable the new test code for header units.

StephanTLavavej commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,355 @@
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using _Mybase   = _Basic_syncbuf_impl<_Elem, _Traits>;
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
```
As far as I can tell, `pointer` and `size_type` aren't depicted as public member types of `basic_syncbuf`. If the implementation needs these, they should be `private`. (Although we commonly uglify internal typedefs, we widely use `pointer`, and it should be okay to do the same for `size_type` - we don't need to worry about macros, we just need to ensure that users can't detect `basic_syncbuf::pointer` etc. through SFINAE.)

cpplearner replied:
I think these names need to be uglified, in case some user-defined class inherits from this class.

```c++
using pointer = void*;
struct mybuf : std::syncbuf {
    pointer p; // name lookup should find `::pointer`, not `syncbuf::pointer`
};
```

StephanTLavavej reviewed (commented):
(Partial review, looked at `<syncstream>` up to `int sync() override`)

MichaelRizkalla commented on stl/inc/ostream:
```diff
@@ -993,6 +993,47 @@ basic_ostream<_Elem, _Traits>& __CLRCALL_OR_CDECL flush(basic_ostream<_Elem, _Tr
+    _Ostr.flush();
+    auto _Sync_buf_ptr = dynamic_cast<_Basic_syncbuf_impl<_Elem, _Traits>*>(_Ostr.rdbuf());
+    if (_Sync_buf_ptr) {
+        if (!_Sync_buf_ptr->_Do_emit()) {
+            _Ostr.setstate(ios::badbit);
```
I did that from cosistnecy point of view, since calling `sb->emit` from `basic_osyncstream` sets `badbit` acording to [syncstream.osyncstream.members]/1: "If that call returns false, calls setstate(ios_base::badbit)".
But I'm doing more than the explicit standard requirement, that's my bad.

MichaelRizkalla commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,355 @@
+_STD_BEGIN
+
+extern "C" shared_mutex* __cdecl _Get_mutex_for_instance(void* _Ptr) noexcept;
+extern "C" void __cdecl _Acquire_mutex_for_instance(void* _Ptr);
```
`_Acquire_mutex_for_instance` is invoking `std::map::try_emplace` and it could fail to allocate.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
```
As this separately compiled source file is being injected into the import lib, we should have the Scary Comment For Import Lib Stuff:
https://github.com/microsoft/STL/blob/47cd70315d560e4fe1ac699c0eb8ec8bfc5cb658/stl/src/vector_algorithms.cpp#L4-L7
(This is moot if we move this into a satellite DLL.)

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// initialize syncstream mutex map
```
We need to determine whether the import lib is the right place for the syncstream mutex map. Because the import lib is effectively statically linked into the user's binary, scenarios such as multiple user binaries (e.g. `meow.dll`, `purr.dll`, `cat.exe`) being linked to a single STL DLL (`msvcp140.dll`) would end up with separate syncstream mutex maps (one per user binary), which seems problematic.

Because these mutexes are global state, I suspect that we need a single copy of them per STL DLL, which means that this needs to be added to a satellite DLL. (For ease of development, the `atomic_wait` DLL recently added for C++20 is probably best to extend, despite the name - `/std:c++latest` is not yet ABI-frozen so we can add exports to that DLL, which will be massively easier than adding a new satellite DLL.)

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+// initialize syncstream mutex map
+
+#include <internal_shared.h>
+#include <map>
```
`map` is an STL container whose representation is affected by `_ITERATOR_DEBUG_LEVEL`. This means that we can't drag it in to the import lib, due to ODR violations. (There are only 2 DLLs/import LIBs for release/debug, yet there are 5 possible settings for IDL.) This is the scenario that the Scary Comment For Import Lib Stuff warns against (with `basic_string` as the example).

If this code is moved into a satellite DLL (as I suspect is necessary), then this issue is moot - satellite DLLs can use STL containers without fear of ODR violations because there is a binary boundary.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+class _Crt_allocator {
+public:
+    using value_type                             = _Ty;
+    using size_type                              = size_t;
+    using difference_type                        = ptrdiff_t;
```
`allocator_traits` provides these as defaults, so these typedefs are unnecessary.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+    constexpr _Crt_allocator(const _Crt_allocator<_Other>&) noexcept {}
+
+    _NODISCARD __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
+        auto _Ptr = _calloc_crt(_Count, sizeof(_Ty));
```
This could be top-level `const`.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+        return static_cast<_Ty*>(_Ptr);
+    }
+
+    void deallocate(_Ty* const _Ptr, const size_t) noexcept {
```
We conventionally don't mark unnamed parameters as `const` (no danger of modifying them :smile_cat:):
```suggestion
    void deallocate(_Ty* const _Ptr, size_t) noexcept {
```

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+    shared_lock _Guard(_Mutex);
+    auto _Instance_mutex_iter = _Mutex_map.find(_Ptr);
+    _ASSERT_EXPR(_Instance_mutex_iter != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _STD addressof(_Instance_mutex_iter->second._Mutex);
```
This is separately compiled (and we control `shared_mutex`) so there's no need to defend against overloaded `operator&`:
```suggestion
    return &_Instance_mutex_iter->second._Mutex;
```

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+    auto _Instance_mutex_iter = _Mutex_map.find(_Ptr);
+    _ASSERT_EXPR(_Instance_mutex_iter != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _STD addressof(_Instance_mutex_iter->second._Mutex);
+}
```
Whitespace nitpick: We conventionally separate function definitions with a newline (except when they are very short and very repetitive). Occurs below.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+}
+extern "C" _CRTIMP2 void __stdcall _Acquire_mutex_for_instance(void* _Ptr) {
+    scoped_lock _Guard(_Mutex);
+    _Mutex_map.try_emplace(_Ptr).first->second._Ref_count++;
```
We conventionally prefer preincrement, which makes any necessary postincrements stand out.

StephanTLavavej replied:
There's a lot going on here, so I'll try adding `auto& _Refs` for clarity.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+}
+extern "C" _CRTIMP2 void __stdcall _Release_mutex_for_instance(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Mutex);
+    auto _Instance_mutex_iter = _Mutex_map.find(_Ptr);
```
This can be top-level `const`.

StephanTLavavej commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,70 @@
+    auto _Instance_mutex_iter = _Mutex_map.find(_Ptr);
+    _ASSERT_EXPR(_Instance_mutex_iter != _Mutex_map.end(), "No mutex exists for given instance!");
+    if (--_Instance_mutex_iter->second._Ref_count == 0) {
+        _Mutex_map.erase(_Ptr);
```
This will perform another O(log N) lookup to find the `_Ptr` key. It would be strictly more efficient to erase `_Instance_mutex_iter`.

StephanTLavavej reviewed (changes requested):
Finishing the previous partial review - the only thing I haven't checked is the latest commit addressing feedback (thanks! I will try to double-check that soon).

StephanTLavavej commented:
I've pushed a merge with `main` (resolving a trivial conflict in the feature-test macros) and small changes to address my second review. Tomorrow I'll work on the remaining issues:

* [ ] reduce/report/workaround the header units compiler bug
* [ ] investigate moving this from the import lib to the satellite DLL
* [ ] deal with `/EHsc` implying `noexcept` for `extern "C"`

This is looking very good and I'm hopeful we can merge this next week (the amount of MS-internal build changes needed should be fairly small, if adding to an existing satellite DLL). Thanks for the awesome work here :heart_eyes_cat:

MichaelRizkalla commented:
I have added a small change that I missed from the first review. Also, I added a *workaround* to `/EHsc` which involves throwing in the header. I made the new function a `template` function to avoid `warning C4505: unreferenced local function has been removed`. Feel free to discard that change if it seems not optimal :cat:

AdamBucior commented on stl/inc/syncstream:
```diff
@@ -0,0 +1,362 @@
+extern "C" void __stdcall _Release_mutex_for_instance(void* _Ptr) noexcept;
+
+template <class _Enabler>
+inline void _Acquire_mutex_for_instance(void* _Ptr) {
```
This function is only called once. It can be inlined into `basic_syncbuf`s constructor.

StephanTLavavej commented:
/azp run

StephanTLavavej commented:
/azp run

StephanTLavavej commented:
I've pushed the following changes:

* Disable the header units test coverage until Microsoft-internal VSO-1275701 "Standard Library Header Units: Runtime crash with dynamic atexit destructor for syncstreams" is fixed.
* In headers, we conventionally call `_Xbad_alloc()` to throw.
* Don't mark the function definitions as `_CRTIMP2`, which is a macro for `__declspec(dllexport)`.
  + When we inject functions into the import lib, they're effectively statically linked, not exported; `filesystem.cpp` is good precedent.
  + This change also applies to the `atomic_wait` satellite DLL, which describes its exports with `msvcp_atomic_wait.src` instead of `__declspec(dllexport)`.
* Rename `_Lookup_map` and `_Lookup_mutex`.
  + After reading and re-reading `syncstream.cpp`, I noticed that it was confusing to have two layers of mutexes, one to protect the map and then an arbitrary number in the nodes to protect each stream. Having them both named `_Mutex` was especially confusing (even though there was no shadowing, strictly speaking). I believe it's clearer to name the map and its corresponding mutex similarly, to indicate that the `_Lookup_map` is protected by the `_Lookup_mutex`.
* Use `size_t` for refcounts instead of `uint64_t`.
  + On 32-bit architectures, we can't possibly have over 4 billion syncbufs.
* Move code into the `atomic_wait` satellite DLL.
  + `stl/CMakeLists.txt`
    - Move `syncstream.cpp` into `SOURCES_SATELLITE_ATOMIC_WAIT`. It needs the `shared_mutex` machinery which is injected into the import lib, so when building the `atomic_wait` DLL, we need to link in the `msvcp${D_SUFFIX}_implib_objects`. The `msvcp_2` satellite was already doing this.
  + `stl/inc/syncstream`
    - Rename the separately compiled functions to follow the `__std_meow` convention.
    - Declare them in an `_EXTERN_C` block outside of `std`.
    - Say `shared_mutex` in their names, because that's the type being accessed.
    - "Flatten" the function to return `void*` instead of a pointer to a C++ Standard type, restored with `static_cast<shared_mutex*>` later. (I am not 100% sure if this is ideal; it's not really a flat C API.)
  + `stl/src/msvcp_atomic_wait.src`
    - Add the new functions to the list of exports.
    - I believe we can do this because this satellite was added for C++20.
  + `stl/src/syncstream.cpp`
    - Include what we use: `<functional>` for `less`, `<mutex>` for `scoped_lock`, `<new>` for `bad_alloc`, `<type_traits>` for `is_same_v` (added), `<utility>` for `pair`.
    - Separate the file into unnamed namespace and `_EXTERN_C` regions (imitating the pattern used elsewhere). This requires `_STD` qualifying stuff.
    - Drop `static` from `_Lookup_map` and `_Lookup_mutex` as they are now within the unnamed namespace.
    - Add a `static_assert` to make it extra clear what type information we're discarding.
* Increase efficiency by caching the mutex pointer.
  + This is a **significant representation change**, please double-check it. I eventually realized that every call to `emit()` was calling `__std_get_shared_mutex_for_instance()`, taking a `shared_lock` (reader lock) on the **global** `_Lookup_mutex`, and holding it while performing an O(log N) node lookup. After releasing that, `emit()` would then acquire the per-streambuf mutex. After thinking carefully, I believe that this is unnecessary. Between the "acquire" and "release" calls, the wrapped streambuf and its corresponding mutex aren't going anywhere; we will always get the same mutex pointer (which is the whole purpose of this machinery; different `basic_syncbuf` wrappers will get the same mutex pointer as long as anything is holding a refcount). So, at the cost of storing an extra pointer in the `basic_syncbuf`, we can simply cache the mutex pointer. This should make `emit()` significantly more efficient, and avoid any global contention. We simply need to be careful to update the cached mutex pointer along with the wrapped streambuf pointer; I believe I got all of the occurrences. (The class invariant is that they are both null or both non-null.)
  + This also simplifies the code by removing the "get" function. Now the "acquire" function returns the mutex pointer. This means that there is no "sharing" going on (all locks acquired are exclusive writer locks), but because `shared_mutex` directly wraps SRWLOCK, I believe it remains optimal to use for both the `_Lookup_mutex` and the per-streambuf mutexes.

StephanTLavavej commented:
I believe the product code is ready, but I realized that I haven't fully reviewed the test code.

MichaelRizkalla commented:
The latest change of adding the `shared_mutex*` member made me realize something:
- Can we use `[[no_unique_address]]` with `allocator_type _Al{};`? I know it's not implemented yet, but I *think* there is a transitional comment.

StephanTLavavej commented:
> Can we use `[[no_unique_address]]` with `allocator_type _Al{};`?

Excellent question, I had forgotten about this. We indeed have a couple of TRANSITION comments, and many commented-out uses in C++20 code (mostly ranges), blocked by #1364.

In this case, I think we can use `_Compressed_pair` (to be replaced by `[[no_unique_address]]` when we get it sorted out, hopefully in vNext). I'll work on changes for that.

StephanTLavavej commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.hpp:
```diff
@@ -0,0 +1,194 @@
+template <class Ty, class Other>
+_NODISCARD bool operator==(const non_move_assignable_non_equal_allocator<Ty>&,
+    const non_move_assignable_non_equal_allocator<Other>&) noexcept {
+    return false;
```
Since this is test-only, I don't think it should block merging.

AdamBucior replied:
Maybe compare the addresses of allocators? Then it would compare equal to itself.

AraHaan replied:
maybe compare address + data + size of the allocator to ensure everything is the same 1st?

StephanTLavavej replied:
Comparing allocator addresses is less non-conformant, but not completely (copies and rebinds must compare equal). Storing an int id would be a complete fix.

MichaelRizkalla replied:
I *think* adding an id will make copies compare to equal but not rebinds. Would it be enough?

StephanTLavavej replied:
If you copy the ID when rebinding from `MyAlloc<X>` to `MyAlloc<Y>`, same as when copy-constructing `MyAlloc<X>` to `MyAlloc<X>`, then you'll satisfy the Standard's guarantees.

StephanTLavavej commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.hpp:
```diff
@@ -0,0 +1,194 @@
+};
+
+template <class Ty>
+class fancy_ptr {
```
This is another test-only issue that I don't think should block merging.

StephanTLavavej replied:
If anyone wants to fix this before merging, `VSO_0000000_fancy_pointers` has a conformant fancy pointer and allocator. I think it would be sufficient to add instantiation coverage to that test (it's compile-only) and remove the coverage here, since the intersection of "fancy pointer and small size allocation" seems like it should work automatically (which is why we're usually satisfied with compile-only coverage for fancy pointers).

MichaelRizkalla replied:
I can do that this morning. I'll remove fancy pointer coverage from here, and add `basic_syncbuf` and `basic_osyncstream` to `VSO_0000000_fancy_pointers`

StephanTLavavej commented:
I reconsidered having the acquire function return `void*`. That doesn't really achieve anything, and it might lead to forgetting the issue in the future. So I've changed the interface to return `shared_mutex*`, with a comment that this should be changed in vNext when we want to truly flatten the DLL's interface (no C++ types anywhere). For v19 (the current ABI), avoiding all C++ types is unnecessary (the STL still exports entire classes from the main DLL); we simply have to avoid any types that depend on `_ITERATOR_DEBUG_LEVEL` (and `shared_mutex` is good there).

AdamBucior commented on stl/src/syncstream.cpp:
```diff
@@ -0,0 +1,83 @@
+#include <mutex>
+#include <new>
+#include <shared_mutex>
+#include <type_traits>
```
`<type_traits>` no longer needs to be included.

StephanTLavavej replied:
`_Crt_allocator` still needs `true_type`. (I was incorrect when I previously mentioned that `<type_traits>` was needed only for the newly-added `is_same_v`.)

AdamBucior commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.hpp:
```diff
@@ -0,0 +1,160 @@
+template <class Ty, class Other>
+[[nodiscard]] bool operator==(const non_move_assignable_non_equal_allocator<Ty>& lhs,
+    const non_move_assignable_non_equal_allocator<Other>& rhs) noexcept {
+    if (lhs.id == rhs.id) {
+        return true;
+    }
+    return false;
```
```suggestion
    return lhs.id == rhs.id;
```

AdamBucior commented on tests/std/tests/P0053R7_cpp_synchronized_buffered_ostream/test.hpp:
```diff
@@ -0,0 +1,160 @@
+private:
+    static size_t id_gen;
+};
+size_t non_move_assignable_non_equal_allocator_id::id_gen = 0;
```
It doesn't really matter because this header is included only once, but this should be `inline`.

StephanTLavavej commented:
Thanks for fixing the fancy pointer and allocator issues! I've pushed two commits, adding `explicit` to the ID class's constructor (it's test code, but I always recommend being cautious about implicitly converting constructors), and adding the necessary `stl/msbuild` changes. I've mirrored this to internal MSVC-PR-303125 along with the other necessary internal changes. :smile_cat: (This is mirrored all by itself, not chained with other GitHub PRs, due to the magnitude of the changes and the importance of the feature.)

Please note: Further changes are still possible (e.g. if you notice something, or if final review leads to more feedback that should be addressed before merging), but they must be replicated to the mirror PR in order to avoid codebase divergence.

StephanTLavavej commented:
/azp run

StephanTLavavej commented:
I've pushed a comment change to an internal `stl/msbuild` file (technically unrelated to this PR, but right above a modification, and @CaseyCarter noticed that the comment was confusing).

StephanTLavavej commented:
I've pushed and mirrored a change to use `_Init_locks`, which resolves the Standard Library Header Units crash (so we can restore that part of the test). There's a Static Initialization Order Fiasco, which was revealed but not inherently caused by modules, where the syncstream `map` requires the STL's locks to be initialized (because for IDL=2, we need to take the debug lock in `map`'s destructor to orphan all of its iterators). If the locks aren't guaranteed to be initialized *before* `map` construction, then they aren't guaranteed to be destroyed *after* `map` destruction, and we're doomed.

The fix is to use the `_Init_locks` code pattern that's used elsewhere in the STL:
https://github.com/microsoft/STL/blob/8f79acf1501fb5feb43d26c4008babd87876839a/stl/src/cout.cpp#L9-L11
This guarantees initialization order within the `syncstream.cpp` translation unit. (I had seen this pattern before, but forgot about it since I've never added occurrences in my 14 years working on the STL - first time for everything.)

I have copied the exact spelling of the `_Init_locks` pattern for searchability, even though we should probably say `_STD` for consistency (that should be a global cleanup). The non-`_Ugly` identifier `initlocks` is fine because this is `static` (again, we could clean it up, but that would be a separate global cleanup).

CaseyCarter reviewed (commented):
Just a couple of `noexcept` questions, looks good otherwise.

StephanTLavavej commented:
Thanks for implementing this new stream! :tada: :rocket: :smiley_cat: This will ship in VS 2019 16.10 Preview 2.
